### ISSUE TITLE
`Unexpected token` when importing `*.d.cts` or `*.d.mts` files

### ISSUE THREAD

**Issue by phun-ky** [ezolenko/rollup-plugin-typescript2#446]
## Troubleshooting

1. Does `tsc` have the same output? If so, please explain why this is incorrect behavior

    N/A

1. Does your Rollup plugin order match [this plugin's compatibility](https://github.com/ezolenko/rollup-plugin-typescript2#compatibility)? If not, please elaborate

    yes

1. Can you create a [minimal example](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces this behavior? Preferably, use [this environment](https://stackblitz.com/edit/rpt2-repro) for your reproduction
    

    https://stackblitz.com/edit/rpt2-repro-3j5vny?file=src%2Findex.ts,package.json,src%2Flib%2Findex.ts,README.md,tsconfig.json

## What happens and why it is incorrect

Error is thrown, should have support for `*.cts`/`*.mts` files

```
[!] (plugin rpt2) RollupError: Unexpected token (Note that you need plugins to import files that are not JavaScript)
../../node_modules/helmet/index.d.cts (3:5)
1: import {IncomingMessage, ServerResponse} from "http"
2: 
3: type ContentSecurityPolicyDirectiveValueFunction = (req: IncomingMessage, res: ServerResponse) => string
        ^
4: type ContentSecurityPolicyDirectiveValue = string | ContentSecurityPolicyDirectiveValueFunction
5: interface ContentSecurityPolicyOptions {

    at error (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:279:30)
    at Module.error (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:13947:16)
    at Module.tryParse (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:14658:25)
    at Module.setSource (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:14260:39)
    at ModuleLoader.addModuleSource (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:24225:20)
```
```
[!] (plugin rpt2) RollupError: Unexpected token (Note that you need plugins to import files that are not JavaScript)
../../node_modules/helmet/index.d.mts (3:5)
1: import {IncomingMessage, ServerResponse} from "http"
2: 
3: type ContentSecurityPolicyDirectiveValueFunction = (req: IncomingMessage, res: ServerResponse) => string
        ^
4: type ContentSecurityPolicyDirectiveValue = string | ContentSecurityPolicyDirectiveValueFunction
5: interface ContentSecurityPolicyOptions {

    at error (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:279:30)
    at Module.error (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:13947:16)
    at Module.tryParse (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:14668:25)
    at Module.setSource (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:14270:39)
    at ModuleLoader.addModuleSource (/home/alexander/Workspace/company/internal/devplatform-server-api/node_modules/rollup/dist/shared/rollup.js:24266:20)
```

## Environment

Ubuntu 22.04

### Versions

```text
  System:
    OS: Linux 5.17 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i7-12700H
    Memory: 43.96 GB / 62.51 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 18.16.0 - /usr/local/bin/node
    npm: 9.5.1 - /usr/local/bin/npm
  npmPackages:
    rollup: ^3.22.0 => 3.22.0 
    rollup-plugin-typescript2: ^0.34.1 => 0.34.1 
    typescript: ^5.0.4 => 5.0.4 
  npmGlobalPackages:
    typescript: 5.0.4

```

<details>
  <summary><h4><code>rollup.config.js</code></h4>: </summary>

```js
/* global process */
import { nodeResolve } from '@rollup/plugin-node-resolve';
import ts from 'rollup-plugin-typescript2';
import dts from 'rollup-plugin-dts';
import json from '@rollup/plugin-json';
import typescript from 'typescript';
import terser from '@rollup/plugin-terser';
import commonjs from '@rollup/plugin-commonjs';
import externals from 'rollup-plugin-node-externals';

const PACKAGE_NAME = process.cwd();

export default [
  {
    input: `${PACKAGE_NAME}/src/index.ts`,
    output: {
      dir: `${PACKAGE_NAME}/dist`,
      format: 'es'
    },
    plugins: [
      externals(),
      json(),
      nodeResolve({ preferBuiltins: false }),
      commonjs(),
      ts({
        useTsconfigDeclarationDir: false,
        sourceMap: false,
        typescript,
        tsconfig: `${PACKAGE_NAME}/tsconfig.json`
      }),
      terser()
    ]
  },
  {
    input: `${PACKAGE_NAME}/dist/index.d.ts`,
    output: [{ file: `${PACKAGE_NAME}/dist/index.d.ts`, format: 'es' }],
    plugins: [dts()]
  }
];

```

</details>

<details>
  <summary><h4><code>tsconfig.json</code></h4>: </summary>

```json5
{
  "compilerOptions": {
    "module": "ES2020",
    "target": "es2022",
    "lib": ["ES2021", "dom", "dom.iterable"],
    "sourceMap": true,
    "declaration": true,
    // "declarationDir": "./dts",
    "moduleResolution": "node",
    "rootDirs": ["middlewares", "utils", "routes"],
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": false,
    "strictNullChecks": true,
    "esModuleInterop": true,
    "typeRoots": ["./@types", "./node_modules/@types"],
    "resolveJsonModule": true,
    "types": ["node"]
  },
  "exclude": ["__tests__", "node_modules", "dist", "*.d.ts", "*.d.cts"]
}

```

</details>

<details>
  <summary><h4><code>package.json</code></h4>: </summary>

```json
{
  "name": "@kilabs/devplatform-server-api",
  "type": "module",
  "private": true,
  "version": "0.0.1",
  "description": "The API monorepo for the devplatform",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "commit": "npx git-cz"
  },
  "author": "Alexander Vassbotn Røyne-Helgesen <alexander.royne-helgesen@company.no>",
  "license": "SEE LICENSE IN LICENSE",
  "devDependencies": {
    "@rollup/plugin-commonjs": "^24.1.0",
    "@rollup/plugin-json": "^6.0.0",
    "@rollup/plugin-node-resolve": "^15.0.2",
    "@rollup/plugin-terser": "^0.4.1",
    "@types/express": "^4.17.17",
    "@typescript-eslint/eslint-plugin": "^5.54.1",
    "@typescript-eslint/parser": "^5.54.1",
    "eslint": "^8.25.0",
    "eslint-config-airbnb": "^19.0.4",
    "eslint-config-prettier": "^8.5.0",
    "eslint-plugin-compat": "^4.1.2",
    "eslint-plugin-import": "^2.27.5",
    "eslint-plugin-prettier": "^4.2.1",
    "lerna": "^6.6.2",
    "prettier": "^2.7.1",
    "prettier-eslint": "^15.0.1",
    "rollup": "^3.22.0",
    "rollup-plugin-dts": "^5.3.0",
    "rollup-plugin-node-externals": "^6.0.1",
    "rollup-plugin-typescript2": "^0.34.1",
    "tslib": "^2.3.1",
    "typescript": "^5.0.4"
  },
  "workspaces": [
    "utils/*",
    "routes/*",
    "middlewares/*"
  ]
}

```

</details>

**Comment by phun-ky** [comment]
Perhaps this is relevant: https://github.com/rollup/plugins/issues/1480

**Comment by phun-ky** [comment]
@agilgur5 well, that kind of goes without saying :) but I am not including the declaration files directly, nor is `helmet`.

**Comment by ezolenko** [comment]
@phun-ky try explicitly including d.cts and d.mts:

`include: [ "*.ts+(|x)", "**/*.ts+(|x)", '**/*.d.cts', '**/*.d.mts' ]`

This will run them through typescript and I guess produce empty js. 

Somehow `import 'helmet'` resolves to .d.cts:
```
rpt2: dependency '/home/projects/rpt2-repro-vnppxj/node_modules/helmet/index.d.cts'
rpt2:     imported by '/home/projects/rpt2-repro-vnppxj/src/index.ts'
rpt2: resolving 'helmet' imported by '/home/projects/rpt2-repro-vnppxj/src/index.ts'
rpt2:     to '/home/projects/rpt2-repro-vnppxj/node_modules/helmet/index.d.cts'
```

**Comment by ezolenko** [comment]
Ah, ignore that -- this fixes the problem: 

```
include: ['**/*.ts', '**/*.cts', '**/*.mts'],
```

**Comment by ezolenko** [comment]
@phun-ky could you try building your whole repo with master branch of the plugin?

**Comment by phun-ky** [comment]
@ezolenko still got issues, sorry for the late reply!

```bash
/home/alexander/Workspace/knowit/internal/devplatform-server-api/packages/middlewares/csp/src/index.ts → dist...
[!] (plugin rpt2) RollupError: Unexpected token (Note that you need plugins to import files that are not JavaScript)
../../../node_modules/helmet/index.d.cts (3:5)
1: import {IncomingMessage, ServerResponse} from "http"
2: 
3: type ContentSecurityPolicyDirectiveValueFunction = (req: IncomingMessage, res: ServerResponse) => string
        ^
4: type ContentSecurityPolicyDirectiveValue = string | ContentSecurityPolicyDirectiveValueFunction
5: interface ContentSecurityPolicyOptions {

    at error (/usr/local/lib/node_modules/rollup/dist/shared/rollup.js:279:30)
    at Module.error (/usr/local/lib/node_modules/rollup/dist/shared/rollup.js:13933:16)
    at Module.tryParse (/usr/local/lib/node_modules/rollup/dist/shared/rollup.js:14644:25)
    at Module.setSource (/usr/local/lib/node_modules/rollup/dist/shared/rollup.js:14246:39)
    at ModuleLoader.addModuleSource (/usr/local/lib/node_modules/rollup/dist/shared/rollup.js:24211:20)

```

**Comment by ezolenko** [comment]
@phun-ky could you post verbose output when building with master branch and your rollup config if different from what you already posted?

**Comment by ezolenko** [comment]
So this run succeeded? Try doing a clean build

**Comment by phun-ky** [comment]
Crap, no, parts of the log is missing

**Comment by phun-ky** [comment]
@ezolenko updated the comment with the output now, sorry

**Comment by phun-ky** [comment]
What's weird is that this works like a charm:

```bash
$ ts-node --esm --project tsconfig.json src/index.ts
``

**Comment by phun-ky** [comment]
Any progress on this issue? :)

**Comment by CHamilton0** [comment]
I'm seeing the same issue as described here when trying to use rollup with this plugin after updating Vuetify from v3.2.x to v3.3.x. I notice that's when they changed the .d.ts imports to .d.mts and that's when the errors started to appear.

**Comment by ezolenko** [comment]
@phun-ky the progress is slow :)

Could you try `mts` branch (set this in your package.json: `"rollup-plugin-typescript2": "git@github.com:ezolenko/rollup-plugin-typescript2.git#mts",`)

Your example project compiles for me on that branch.

**Comment by phun-ky** [comment]
@ezolenko yep! that worked like a charm!

**Comment by ezolenko** [comment]
Ok, I'll make a release with the last few fixes

**Comment by ezolenko** [comment]
In 0.35.0 now

**Comment by phun-ky** [comment]
@ezolenko thanks for the great work! I really appreciate it!

**Comment by agilgur5** [comment]
> @agilgur5 well, that kind of goes without saying :) but I am not including the declaration files directly, nor is `helmet`.

Ah, sorry for not checking this in more depth earlier! Was on my phone and just responded real quick to the title. We've had a lot of folks trying to import `.d.ts` files in the past so thought that a quick response with a link to a similar issue might be helpful.

Glad that ezolenko could had some time to spare and could help in more depth that same day!

## root cause analysis 

> Somehow `import 'helmet'` resolves to .d.cts:

So I did root cause this as I was curious how the heck that happened as that's definitely not correct behavior. Took me a bit to find Helmet's [`package.json`](https://github.com/helmetjs/helmet/blob/95bd2cc6045551048b8af305702996982d6ea7eb/build/build-package.ts#L203) as it is dynamically generated. Its current version's internal only have `.cjs`, `.mjs`, `.cts` extensions, which rpt2 did not previously support.
Specifically, rpt2 resolved one of the references to the [`index.d.cts` file](https://github.com/helmetjs/helmet/blob/95bd2cc6045551048b8af305702996982d6ea7eb/build/build-package.ts#L230), which is a possible correct resolution, but rpt2 is _supposed to_ ignore declarations. Those checks detected declarations as files with `.d.ts`, so ezolenko had to change [this line](https://github.com/ezolenko/rollup-plugin-typescript2/pull/451/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L277) and our default `exclude` filter to now _also_ ignore `.d.cts` and `.d.mts` as those are now possible extensions in newer TS versions.

That is to say, the `Unexpected token` error is actually correct, as rpt2 shouldn't be reading declarations, but it shouldn't be loading declarations into Rollup at all to begin with. It just didn't know the new file extensions in order to ignore them. Will rename the issue to make that clearer.

### irony of incomplete branches

In a bout of extreme irony, when I looked at my local code, I actually already had a similar line there 😵‍💫🤔 It turns out while I was attempting to fix #426 _back in October_ (never quite finished that), I made [this commit](https://github.com/agilgur5/rollup-plugin-typescript2/commit/61ef625b85dda9f397fecc0f17dfe482d4927700) on one of my branches containing the exact same line 😅
